sql: fix and enhance expression normalization.

knz opened the pull request:
This patch is needed so that the expressions in the Sqllogictest do
not cause the server to panic.

Summary:
- Bug fixes:
  - an invalid cast was attempted for some expressions; is now fixed.
  - NULL caused invalid transformations, NULL is now normalized on a
    fast path.
  - transformations that rebalance arithmetic across a comparison now
    first check that the proposed replacement arithmetic expression
    has a valid overload w.r.t typing.
- Improvements: more simplifications.

Fixes #7412.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7512)

<!-- Reviewable:end -->

knz committed:
sql: fix and enhance expression normalization.

This patch is needed so that the expressions in the Sqllogictest do
not cause the server to panic.

Summary:

- Bug fixes:

  - an invalid cast was attempted for some expressions; is now fixed.
  - NULL caused invalid transformations, NULL is now normalized on a
    fast path.
  - transformations that rebalance arithmetic across a comparison now
    first check that the proposed replacement arithmetic expression
    has a valid overload w.r.t typing.

- Improvements: more simplifications.

knz commented:
cc @petermattis @nvanbenschoten.

tamird commented:
Reviewed 2 of 4 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/parser/col_types.go, line 275 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrF3VOhiF7t8eAbsd:-KLNrF3Wmp0SMfugnNcl:1177599082) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/col_types.go#L275)):*
> ```Go
> 		return &BytesColType{"BYTES"}
> 	}
> 	panic("unknown type")
> ```

log the type

---

*[sql/parser/normalize.go, line 611 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrZ3LvsjOBJ3ui21N:-KLNrZ3LvsjOBJ3ui21O:296275497) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/normalize.go#L611)):*
> ```Go
> }
> 
> // CanBeNumericZero returns true if the expr may be a number and
> ```

i don't think that's what it does...

---

*[sql/parser/normalize.go, line 654 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrfbMhXzMP2G3DQIT:-KLNrfbMhXzMP2G3DQIU:1177599082) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/normalize.go#L654)):*
> ```Go
> 		return NewDInt(0)
> 	}
> 	panic("zero not defined here")
> ```

log the type

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/parser/col_types.go, line 275 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrF3VOhiF7t8eAbsd:-KLRFWz3yMQpEKdy5vV0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/col_types.go#L275)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> log the type
</details>
Done.

---

*[sql/parser/normalize.go, line 611 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrZ3LvsjOBJ3ui21N:-KLRFX_G89RRXVUjttm5:662068717) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/normalize.go#L611)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i don't think that's what it does...
</details>
This is what it should do. Is there a bug in the code?

---

*[sql/parser/normalize.go, line 654 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrfbMhXzMP2G3DQIT:-KLRFapPhpgfVCNb8A3D:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/normalize.go#L654)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> log the type
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/parser/col_types.go, line 275 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrF3VOhiF7t8eAbsd:-KLSpb2mVt6KR3yOW5y9:-469765787) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/col_types.go#L275)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Done.
</details>
ditto

---

*[sql/parser/normalize.go, line 611 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrZ3LvsjOBJ3ui21N:-KLSpJP107qQBCXrQAU5:1297702634) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/normalize.go#L611)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> This is what it should do. Is there a bug in the code?
</details>
It currently returns true if the argument is e.g. a `*DString`

---

*[sql/parser/normalize.go, line 654 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrfbMhXzMP2G3DQIT:-KLSpQyFSuYy6UH7YI0K:1770807632) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/normalize.go#L654)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Done.
</details>
This convention is weak, but we use `log.Fatalf` to avoid the double call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/parser/col_types.go, line 275 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrF3VOhiF7t8eAbsd:-KLSyhwwE94LB4yt6XLx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/col_types.go#L275)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[sql/parser/normalize.go, line 611 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrZ3LvsjOBJ3ui21N:-KLT1pY5BCOwuausFr7B:1235104874) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/normalize.go#L611)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It currently returns true if the argument is e.g. a `*DString`
</details>
Aha, I see. I have extended the logic and also the explanatory comment accordingly. Thanks for spotting this.

---

*[sql/parser/normalize.go, line 654 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLNrfbMhXzMP2G3DQIT:-KLSyjI7p3axEKGs2RNe:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/662245b46527f0299a65ce052d00a4e13b5705a1/sql/parser/normalize.go#L654)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> This convention is weak, but we use `log.Fatalf` to avoid the double call.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/parser/normalize.go, line 676 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLTAmNT9r0TxoYsANUX:-KLTAmNT9r0TxoYsANUY:-740903468) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e2f7fcdaeae7e655951e62c6072d54fdcad60b/sql/parser/normalize.go#L676)):*
> ```Go
> 		}
> 		if _, ok := d.(dividerDatum); ok {
> 			log.Errorf("bug: unknown dividerDatum in IsValidDivider(): %T", d)
> ```

oh good god. how about an error return?

---

*[sql/parser/normalize.go, line 715 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLT3KPcq7Buo13kLyuI:-KLTA_7NLGG-lb-RgYPw:1092091685) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e2f7fcdaeae7e655951e62c6072d54fdcad60b/sql/parser/normalize.go#L715)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> @tamird hmm are you ok with having to explicitly write "return" after a call to log.Fatalf? Can we inform the compiler this does not return?
</details>
Yeah, that's a bummer. I don't know if we can.

By the way, I'm pretty grumpy about these panics, since future changes (like adding a type) will suddenly open up a myriad ways of crashing the server. How about we bite the bullet and plumb errors through? Can be a followup.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/parser/normalize.go, line 676 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLTAmNT9r0TxoYsANUX:-KLTEDnvq4rLHn9sLONM:-1708647443) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e2f7fcdaeae7e655951e62c6072d54fdcad60b/sql/parser/normalize.go#L676)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oh good god. how about an error return?
</details>
But that's the thing here. Normalization is just an optimization, there is no reason to cause the (likely valid) user queries to fail with an error although we didn't finish the job here properly.

---

*[sql/parser/normalize.go, line 715 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLT3KPcq7Buo13kLyuI:-KLTEPMmiYgyjBjfVlpX:91208771) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e2f7fcdaeae7e655951e62c6072d54fdcad60b/sql/parser/normalize.go#L715)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yeah, that's a bummer. I don't know if we can.
> 
> By the way, I'm pretty grumpy about these panics, since future changes (like adding a type) will suddenly open up a myriad ways of crashing the server. How about we bite the bullet and plumb errors through? Can be a followup.
</details>
Ok perhaps there is wisdom for this one, since cannot complete the work safely at this point. What is the new best way to signal an "internal" error not caused by a problem in the user's query?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/parser/normalize.go, line 715 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLT3KPcq7Buo13kLyuI:-KLTFXBtIoOufGndUlR2:2086297948) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e2f7fcdaeae7e655951e62c6072d54fdcad60b/sql/parser/normalize.go#L715)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Ok perhaps there is wisdom for this one, since cannot complete the work safely at this point. What is the new best way to signal an "internal" error not caused by a problem in the user's query?
</details>
just return an error =/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/parser/normalize.go, line 676 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLTAmNT9r0TxoYsANUX:-KLWqeMioS7KraW8kqBA:-2035924521) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e2f7fcdaeae7e655951e62c6072d54fdcad60b/sql/parser/normalize.go#L676)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> But that's the thing here. Normalization is just an optimization, there is no reason to cause the (likely valid) user queries to fail with an error although we didn't finish the job here properly.
</details>
I added the error as discussed on the mailing list.

---

*[sql/parser/normalize.go, line 715 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLT3KPcq7Buo13kLyuI:-KLWqhG47a-WWN7lc1cT:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/05e2f7fcdaeae7e655951e62c6072d54fdcad60b/sql/parser/normalize.go#L715)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> just return an error =/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/normalize.go, line 705 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLWsZKpu-4Sy62XTKus:-KLWsZKqNOabCnZkH8b_:-5011416) ([raw file](https://github.com/cockroachdb/cockroach/blob/50ffd0f50aac3f500116fa273e835deb8f7f813e/sql/parser/normalize.go#L705)):*
> ```Go
> 		}
> 		if _, ok := d.(dividerDatum); ok {
> 			return true, errors.Errorf("internal error: unknown dividerDatum in IsValidDivider(): %T", d)
> ```

what is `IsValidDivider()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/normalize.go, line 705 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLWsZKpu-4Sy62XTKus:-KLWt_vC5DWmDObME7qZ:1332827296) ([raw file](https://github.com/cockroachdb/cockroach/blob/50ffd0f50aac3f500116fa273e835deb8f7f813e/sql/parser/normalize.go#L705)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what is `IsValidDivider()`?
</details>
It's the name of the function, obviously. Why do you ask?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/normalize.go, line 705 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KLWsZKpu-4Sy62XTKus:-KLWu8ouhRQdD5ai72bL:101591285) ([raw file](https://github.com/cockroachdb/cockroach/blob/50ffd0f50aac3f500116fa273e835deb8f7f813e/sql/parser/normalize.go#L705)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> It's the name of the function, obviously. Why do you ask?
</details>
For a friend.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Do you mind running a SQL benchmark diff over this change, just to make sure there weren't any exceptionally inefficient normalization steps introduced. 

---

Reviewed 1 of 2 files at r6, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[sql/parser/col_types.go, line 259 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM13HM8R7De_JxXKBaU:-KM13HM8R7De_JxXKBaV:-1709467905) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/col_types.go#L259)):*
> ```Go
> 	switch d.(type) {
> 	case *DInt:
> 		return &IntColType{"INT", 0}, nil
> ```

Don't we already have all of these defined (see `intColTypeInt`, `floatColTypeFloat`, etc.)? If the result of this function is always immutable, they could be returned to avoid allocations and duplicated logic.

---

*[sql/parser/expr.go, line 994 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM13oU-HTaoN8k81aCJ:-KM13oU-HTaoN8k81aCK:832082028) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/expr.go#L994)):*
> ```Go
> }
> 
> // newBinExprIfValidOverloads constructs a new BinaryExpr if and only
> ```

`s/newBinExprIfValidOverloads/newBinExprIfValidOverload/`

---

*[sql/parser/expr.go, line 995 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM13rnA8UZUNPfdxDgr:-KM13rnA8UZUNPfdxDgs:565695143) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/expr.go#L995)):*
> ```Go
> 
> // newBinExprIfValidOverloads constructs a new BinaryExpr if and only
> // if the pair of arguments have a valid implementation.
> ```

`s/implementation/implementation for the given BinaryOperator/`

---

*[sql/parser/expr.go, line 1006 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM147KzfWBKL247VZi6:-KM147KzfWBKL247VZi7:675034309) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/expr.go#L1006)):*
> ```Go
> 			fn:       fn,
> 		}
> 		expr.typ = fn.ReturnType
> ```

nit: We usually use the `overloadImpl.returnType()` method for this use case

---

*[sql/parser/normalize.go, line 34 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM15CBpv6sSfwV-x6a8:-KM15CBpv6sSfwV-x6a9:788165834) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L34)):*
> ```Go
> 
> func (expr *CoalesceExpr) normalize(v *normalizeVisitor) TypedExpr {
> 	for i := range expr.Exprs {
> ```

I like this normalization, but it's a bit more involved than the others. It could use a short comment.

---

*[sql/parser/normalize.go, line 37 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM15P1AG1tk-UCPqQ2z:-KM15P1AG1tk-UCPqQ3-:-623716764) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L37)):*
> ```Go
> 		subExpr := expr.TypedExprAt(i)
> 
> 		if i == len(expr.Exprs)-1 {
> ```

nit: Something like `if last := (i == len(expr.Exprs)-1); last {` can often be a harmless way to improve skim-ability.

---

*[sql/parser/normalize.go, line 67 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM16FevI6xa-5gTdMiL:-KM16Fewk929NwIS4g0k:72907471) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L67)):*
> ```Go
> 			return expr
> 		}
> 		if d, err := GetBool(cond.(Datum)); err == nil {
> ```

Doesn't `.Eval()` return a `Datum`?

---

*[sql/parser/normalize.go, line 95 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM19MF6VlUm5eMxylny:-KM19MF6VlUm5eMxylnz:-1085507917) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L95)):*
> ```Go
> 		}
> 		switch b := val.(type) {
> 		// -(a - b) -> (b - a)
> ```

Double checking: This property is true of all data types, right?

---

*[sql/parser/normalize.go, line 162 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM19ycSXfCoBgLiqEMa:-KM19ycSXfCoBgLiqEMb:364030902) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L162)):*
> ```Go
> 			cbz, err := CanBeZeroDivider(right)
> 			if err != nil {
> 				final, v.err = expr, err
> ```

We don't need `final = expr` in these situations.

---

*[sql/parser/normalize.go, line 228 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1A4aDL6ntyhM9FTuN:-KM1A4aE2L6yWmK6JYKQ:1579396839) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L228)):*
> ```Go
> 		}
> 		if right != DNull {
> 			if d, err := GetBool(right.(Datum)); err == nil {
> ```

same as above, isn't `right` already a `Datum`

---

*[sql/parser/normalize.go, line 489 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1B0PQMJToGDCfl7eQ:-KM1B0PQMJToGDCfl7eR:-1336667767) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L489)):*
> ```Go
> func (expr *ParenExpr) normalize(v *normalizeVisitor) TypedExpr {
> 	newExpr := expr.TypedInnerExpr()
> 	if normalizeable, ok := newExpr.(normalizableExpr); ok {
> ```

I was recently thinking about this. Does the walk infrastructure/normalizeVisitor take care of recursing into modified returned expressions?

---

*[sql/parser/normalize.go, line 621 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1Bp2LRzZOYFkCMm6a:-KM1Bp2Mp3Ar0G43O6Bq:-1134718848) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L621)):*
> ```Go
> 		return GT, nil
> 	default:
> 		return op, errors.Errorf("internal error: unable to invert: %s", op)
> ```

If this is a common case, then allocating a new error every time is going to be an issue.

---

*[sql/parser/normalize.go, line 728 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1COdVp2dvspjV2dy3:-KM1COdVp2dvspjV2dy4:-2027074007) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L728)):*
> ```Go
> 			return *t == 0, nil
> 		}
> 		if _, ok := d.(dividerDatum); ok {
> ```

Can this be added to the type switch?

---

*[sql/parser/type_check.go, line 58 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM18Kmw1V42Dn6G-yUN:-KM18Kmw1V42Dn6G-yUO:-868717954) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/type_check.go#L58)):*
> ```Go
> // are added.
> type dividerDatum interface {
> 	isDividerDatum()
> ```

nit: This doesn't seem to fit our naming convention for "annotation methods". It should probably just be `dividerDatum()`.

---

*[sql/parser/type_check.go, line 61 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM16jqbrANds-4lR2D_:-KM16jqbrANds-4lR2Da:548558213) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/type_check.go#L61)):*
> ```Go
> }
> 
> var _ dividerDatum = NewDInt(0)
> ```

Can these be changed to `var _ dividerDatum = TypeInt`, etc.?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-:-KM1CqbUJEiQdPC8Kmnu:-743443565)*
<!-- Sent from Reviewable.io -->

knz commented:
Running the benchmarks now. Will post the results later.

---

Review status: 1 of 5 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[sql/parser/col_types.go, line 259 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM13HM8R7De_JxXKBaU:-KMA57F5_dS3tHw7fGq4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/col_types.go#L259)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Don't we already have all of these defined (see `intColTypeInt`, `floatColTypeFloat`, etc.)? If the result of this function is always immutable, they could be returned to avoid allocations and duplicated logic.
</details>
Done.

---

*[sql/parser/expr.go, line 994 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM13oU-HTaoN8k81aCJ:-KMA5F6yJZ5mXnv0mZRH:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/expr.go#L994)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/newBinExprIfValidOverloads/newBinExprIfValidOverload/`
</details>
Done.

---

*[sql/parser/expr.go, line 995 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM13rnA8UZUNPfdxDgr:-KMA5IDNBNj_hUUyN9Zi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/expr.go#L995)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/implementation/implementation for the given BinaryOperator/`
</details>
Done.

---

*[sql/parser/expr.go, line 1006 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM147KzfWBKL247VZi6:-KMA5R_718zku9yf285C:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/expr.go#L1006)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nit: We usually use the `overloadImpl.returnType()` method for this use case
</details>
Done.

---

*[sql/parser/normalize.go, line 34 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM15CBpv6sSfwV-x6a8:-KMA6QneB6RowE6Bb_Uj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L34)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I like this normalization, but it's a bit more involved than the others. It could use a short comment.
</details>
Done.

---

*[sql/parser/normalize.go, line 37 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM15P1AG1tk-UCPqQ2z:-KMA6WB0zCaHQgyNnaS8:-2034429587) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L37)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nit: Something like `if last := (i == len(expr.Exprs)-1); last {` can often be a harmless way to improve skim-ability.
</details>
Nice! I wasn't aware of this idiom. 

---

*[sql/parser/normalize.go, line 67 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM16FevI6xa-5gTdMiL:-KMA7aVSahUlr23OCusB:974950386) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L67)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Doesn't `.Eval()` return a `Datum`?
</details>
It does! Thanks for the hint.

---

*[sql/parser/normalize.go, line 95 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM19MF6VlUm5eMxylny:-KMA8xQF5U4Clv1BuM6B:535566334) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L95)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Double checking: This property is true of all data types, right?
</details>
Woops, I checked and indeed it's not guaranteed. I added a conditional for this.

---

*[sql/parser/normalize.go, line 162 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM19ycSXfCoBgLiqEMa:-KMA9H6l1f4Gy5bkbUPN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L162)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> We don't need `final = expr` in these situations.
</details>
Done.

---

*[sql/parser/normalize.go, line 228 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1A4aDL6ntyhM9FTuN:-KMA9IFkFWLoJFLf6cf8:997773187) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L228)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> same as above, isn't `right` already a `Datum`
</details>
It was to avoid an extra temporary. I added it.

---

*[sql/parser/normalize.go, line 489 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1B0PQMJToGDCfl7eQ:-KMA9LuK_R8uDgE21ThY:-166699569) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L489)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I was recently thinking about this. Does the walk infrastructure/normalizeVisitor take care of recursing into modified returned expressions?
</details>
It doesn't; but neither should it need to. The logic where it would make a difference is already inlined in the normalize function (for ComparisonExpr mostly). For the other cases it would result in duplicated work. 

Note how I did have to insert the extra related necessary calls for normalize of RangeCond.

---

*[sql/parser/normalize.go, line 621 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1Bp2LRzZOYFkCMm6a:-KMA9haH57ZH8iv5KUrC:-1539983603) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L621)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> If this is a common case, then allocating a new error every time is going to be an issue.
</details>
There was a panic() here before, so it's not common.

---

*[sql/parser/normalize.go, line 728 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1COdVp2dvspjV2dy3:-KMAAB-neXrT9c37rqSw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L728)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Can this be added to the type switch?
</details>
Done.

---

*[sql/parser/type_check.go, line 58 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM18Kmw1V42Dn6G-yUN:-KMA9vBAfBdodK-H0fRo:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/type_check.go#L58)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> nit: This doesn't seem to fit our naming convention for "annotation methods". It should probably just be `dividerDatum()`.
</details>
Done.

---

*[sql/parser/type_check.go, line 61 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM16jqbrANds-4lR2D_:-KMAA-GHgzbyCETzP3iK:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/type_check.go#L61)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Can these be changed to `var _ dividerDatum = TypeInt`, etc.?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-:-KMAAJuNhwEWEM5j2TKs:1217225120)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 1 of 5 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[sql/parser/type_check.go, line 61 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM16jqbrANds-4lR2D_:-KMAJ4nF0JLZlRZ2Ljyt:1923329173) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/type_check.go#L61)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Done.
</details>
Scrub that; no it can't. TypeXXX instances are typed as Datum, here we want the particular types.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

knz commented:
So the only benchmark that has expressions which would test something related are Select2 and Select3. This is what I measure:

```
name                 old time/op  new time/op  delta
Select2_Cockroach-8   467µs ± 1%   469µs ± 1%  +0.32%  (p=0.031 n=19+18)
Select3_Cockroach-8   611µs ± 1%   612µs ± 1%    ~     (p=0.461 n=19+20)
```

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Besides a few comments.

---

Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[sql/parser/expr.go, line 996 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KMAry9EiA5a3K2-SLQ7:-KMAry9EiA5a3K2-SLQ8:384409607) ([raw file](https://github.com/cockroachdb/cockroach/blob/6913ca180f0cdb1776c61d3f560f811ad6a2eb51/sql/parser/expr.go#L996)):*
> ```Go
> // newBinExprIfValidOverload constructs a new BinaryExpr if and only
> // if the pair of arguments have a valid implementation for the given
> // BinaryOperator..
> ```

`s/.././`

---

*[sql/parser/normalize.go, line 728 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1COdVp2dvspjV2dy3:-KMAt2AojKhESYoG55af:866735676) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L728)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Done.
</details>
Even better, does `case dividerDatum:` work?

---

*[sql/parser/normalize.go, line 113 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KMAsR8zxklLfGDXe750:-KMAsR8zxklLfGDXe751:-173432005) ([raw file](https://github.com/cockroachdb/cockroach/blob/6913ca180f0cdb1776c61d3f560f811ad6a2eb51/sql/parser/normalize.go#L113)):*
> ```Go
> 				return b
> 			}
> 		// - (- a) -> a
> ```

Same question here actually. Is this true for all datum types?

---

*[sql/parser/type_check.go, line 61 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM16jqbrANds-4lR2D_:-KMAtBIlnOhskDR23ott:-915498238) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/type_check.go#L61)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Scrub that; no it can't. TypeXXX instances are typed as Datum, here we want the particular types.
</details>
Ahh ok that's too bad.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-:-KMAtFGXL-CqeUROMUe5:522216893)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[sql/parser/expr.go, line 996 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KMAry9EiA5a3K2-SLQ7:-KMB2yJ4JmOMfREisotx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6913ca180f0cdb1776c61d3f560f811ad6a2eb51/sql/parser/expr.go#L996)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/.././`
</details>
Done.

---

*[sql/parser/normalize.go, line 728 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1COdVp2dvspjV2dy3:-KMB3JQZtVTFz0Ab5nWL:-1039014017) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L728)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Even better, does `case dividerDatum:` work?
</details>
Apparently it does, but somehow it feels a bit uncomfortable: the dividerDatum case is inclusive of the cases before. Are you sure that would be an improvement?

---

*[sql/parser/normalize.go, line 113 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KMAsR8zxklLfGDXe750:-KMB2zF2vAbo5p335EPH:-447942587) ([raw file](https://github.com/cockroachdb/cockroach/blob/6913ca180f0cdb1776c61d3f560f811ad6a2eb51/sql/parser/normalize.go#L113)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Same question here actually. Is this true for all datum types?
</details>
Yes that is true.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 1 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[sql/parser/normalize.go, line 728 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1COdVp2dvspjV2dy3:-KMBNWXJNzfP1Iao7ZD3:-1802377731) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L728)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Apparently it does, but somehow it feels a bit uncomfortable: the dividerDatum case is inclusive of the cases before. Are you sure that would be an improvement?
</details>
That seems the most natural to me. It's basically a restricted default case, which is exactly what we want and definitely an improvement. You could add a `// Catch unhandled dividerDatum impl types` comment if you want.

The only other thing that might be clearer would be to have another type assertion to `dividerDatum` above (which wraps this entire type switch), and use a default case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[sql/parser/normalize.go, line 728 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-KM1COdVp2dvspjV2dy3:-KMETIIi8TN4Y4gjAKRr:1518480136) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0c916c14f587233aecfe0b18680ac6dbc108bb/sql/parser/normalize.go#L728)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> That seems the most natural to me. It's basically a restricted default case, which is exactly what we want and definitely an improvement. You could add a `// Catch unhandled dividerDatum impl types` comment if you want.
> 
> The only other thing that might be clearer would be to have another type assertion to `dividerDatum` above (which wraps this entire type switch), and use a default case.
</details>
Ah I like your last proposal. I just implemented that, PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
LGerTM

---

Reviewed 1 of 1 files at r10, 3 of 3 files at r12.
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7512#-:-KMG0oiivpXZsv61fwx0:1898180902)*
<!-- Sent from Reviewable.io -->
